### PR TITLE
tests: postgres consistency framework

### PR DESCRIPTION
**Pull request by nrainer-materialize**
### Motivation

This PR generalizes the output consistency test framework to allow comparisons against Postgres.

### For reviewers
* Commit-wise review strongly recommended.
* Very sorry for the the size of the PR. It got too large 😢

### CODE REVIEW

File: ci/nightly/pipeline.yml [hunk 1]
```diff
@@ -674,6 +675,17 @@ steps:
           composition: output-consistency
           args: ["--seed=$BUILDKITE_JOB_ID", "--max-runtime-in-sec=1200"]
 
+  - id: postgres-consistency-test
+    label: "Postgres consistency test"
```
Comment (philip-stoev) [inline ci/nightly/pipeline.yml:679]: You can consider renaming this to "Output consistency test (Postgres)" or somehow indicate that this and the previous Buildkite step belong to the same family.

It may also be a good idea to have the two ids start with the same prefix, that is, name this step `output-consistency-test-postgres` or something along those lines.

File: test/postgres-consistency/README.md [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# Postgres consistency tests
+
+## Overview
```
Comment (philip-stoev) [inline test/postgres-consistency/README.md:3]: Can you put some pointers to documentation that describes how the queries are generated, etc?

Alternatively, fold this workflow into `test/output-consistency` if there is enough commonality and extend the `README.md` in that directory.

File: misc/python/materialize/postgres_consistency/execution/pg_dialect_adjuster.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        return type_name
+
+    def adjust_value(self, string_value: str, type_name: str) -> str:
+        if type_name == "INTERVAL":
+            # cut milliseconds away
+            string_value = re.sub("\\.\\d+'$", "'", string_value)
+
+        if type_name.startswith("INT") and string_value.startswith("-"):
+            # wrap negative numbers in parentheses
+            string_value = f"({string_value})"
```
Comment (philip-stoev) [inline misc/python/materialize/postgres_consistency/execution/pg_dialect_adjuster.py:30]: Why did you need to add this exception? If we are parsing negative numbers differently from Pg that would be a Mz product bug.

File: misc/python/materialize/postgres_consistency/ignore_filter/pg_inconsistency_ignore_filter.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+class PgPreExecutionInconsistencyIgnoreFilter(
+    PreExecutionInconsistencyIgnoreFilterBase
+):
+    def _matches_problematic_function_invocation(
+        self,
+        db_function: DbFunction,
+        expression: ExpressionWithArgs,
+        all_involved_characteristics: set[ExpressionCharacteristics],
+    ) -> IgnoreVerdict:
+        inconsistent_value_ordering_functions = ["array_agg", "string_agg"]
```
Comment (philip-stoev) [inline misc/python/materialize/postgres_consistency/ignore_filter/pg_inconsistency_ignore_filter.py:58]: JFYI `ORDER BY` is supported as an argument to those functions.

**Review by philip-stoev — APPROVED** [review]
Thank you! Does it currently run to completion without any failures? If so, in the CI it goes!

**Comment by nrainer-materialize** [comment]
> Thank you! Does it currently run to completion without any failures? If so, in the CI it goes!

I will still need some fine-tuning regarding ignore entries.

File: misc/python/materialize/postgres_consistency/execution/pg_dialect_adjuster.py (hunk 3 above)
Comment (nrainer-materialize) [reply misc/python/materialize/postgres_consistency/execution/pg_dialect_adjuster.py:30]: Yes, I will file an issue.

The evaluation order is different between Postgres and mz. This results in an out-of-range error in Postgres but does not in mz.

Postgres evaluates `SELECT -2147483648::INT` as `SELECT -(2147483648::INT)` while mz does `SELECT (-2147483648)::INT`. `(2147483648::INT)` is out of integer range.

File: ci/nightly/pipeline.yml (hunk 1 above)
Comment (nrainer-materialize) [reply ci/nightly/pipeline.yml:679]: Will do

Comment (nrainer-materialize) [reply ci/nightly/pipeline.yml]: Changed the Buildkite ids and names in https://github.com/MaterializeInc/materialize/pull/21781/commits/049582dfbe1b751902e7703d47af98dc9db81057.
Let me know if you would like me to rename the folders as well.

File: test/postgres-consistency/README.md (hunk 2 above)
Comment (nrainer-materialize) [reply test/postgres-consistency/README.md:3]: ✅ Addressed in https://github.com/MaterializeInc/materialize/pull/21781/commits/4f71e7b234e3505ae43ee43b81e08af28dd72539.

File: misc/python/materialize/postgres_consistency/ignore_filter/pg_inconsistency_ignore_filter.py (hunk 4 above)
Comment (nrainer-materialize) [reply misc/python/materialize/postgres_consistency/ignore_filter/pg_inconsistency_ignore_filter.py:58]: ✅ Tightened the ignore pattern.

File: misc/python/materialize/postgres_consistency/execution/pg_dialect_adjuster.py (hunk 3 above)
Comment (nrainer-materialize) [reply misc/python/materialize/postgres_consistency/execution/pg_dialect_adjuster.py:30]: ✅ Filed https://github.com/MaterializeInc/materialize/issues/21993.

**Comment by nrainer-materialize** [comment]
Nightlies: https://buildkite.com/materialize/nightlies/builds?branch=nrainer-materialize%3Afeature%2Fpostgres-consistency
